src/btree/bt_cursor.c

```diff
@@ -232,107 +232,197 @@ __cursor_fix_implicit(WT_BTREE *btree, WT_CURSOR_BTREE *cbt)
 }
 
 /*
- * __wt_cursor_valid --
- *     Return if the cursor references an valid key/value pair.
+ * __cursor_valid_insert --
+ *     Check the insert list for a valid update.
  */
-int
-__wt_cursor_valid(
-  WT_CURSOR_BTREE *cbt, WT_ITEM *key, uint64_t recno, bool *valid, bool check_bounds)
+static int
+__cursor_valid_insert(WT_CURSOR_BTREE *cbt, WT_ITEM *key, bool *valid, bool check_bounds)
 {
-    WT_BTREE *btree;
-    WT_CELL *cell;
-    WT_COL *cip;
     WT_ITEM tmp_key;
+    WT_SESSION_IMPL *session;
+    bool key_out_of_bounds;
+
+    key_out_of_bounds = false;
+    session = CUR2S(cbt);
+
+    if (cbt->ins == NULL)
+        return (0);
+
+    if (check_bounds && WT_CURSOR_BOUNDS_SET(&cbt->iface)) {
+        /* Get the insert list key. */
+        if (key == NULL && CUR2BT(cbt)->type == BTREE_ROW) {
+            tmp_key.data = WT_INSERT_KEY(cbt->ins);
+            tmp_key.size = WT_INSERT_KEY_SIZE(cbt->ins);
+            WT_RET(__btcur_bounds_contains_key(
+              session, &cbt->iface, &tmp_key, WT_RECNO_OOB, &key_out_of_bounds, NULL));
+        } else
+            WT_RET(__btcur_bounds_contains_key(
+              session, &cbt->iface, key, cbt->recno, &key_out_of_bounds, NULL));
+
+        /* The key value pair we were trying to return weren't within the given bounds. */
+        if (key_out_of_bounds)
+            return (0);
+    }
+
+    WT_RET(__wt_txn_read_upd_list(session, cbt, cbt->ins->upd));
+    if (cbt->upd_value->type != WT_UPDATE_INVALID) {
+        if (cbt->upd_value->type == WT_UPDATE_TOMBSTONE)
+            return (0);
+        *valid = true;
+    }
+
+    return (0);
+}
+
+/*
+ * __cursor_valid_ondisk --
+ *     Check the ondisk value or history store for a valid update.
+ */
+static int
+__cursor_valid_ondisk(WT_CURSOR_BTREE *cbt, WT_ITEM *key, WT_UPDATE *upd, bool *valid)
+{
+    WT_SESSION_IMPL *session;
+
+    session = CUR2S(cbt);
+
+    /*
+     * Check for a value on disk or in the history store, passing in any update.
```

Comment:
Sometimes we say `ondisk value`, sometimes `value on disk`, I also found in the code `on-disk value`, we should probably agree on one terminology.

Reply:
I agree there isn't a standard way to refer to this, I think we can discuss it with the wider team in the retro. I'll make the comments in this PR 'on-disk value' as that is my preferred terminology.

Reply:
f35268fc8